### ISSUE TITLE
Gaps in Zoom/Sizing Controls

### ISSUE THREAD

**Issue by vassudanagunta** [jurplel/qView#460]
First let me say I highly appreciate qView and the work that went into it, especially since it is free and open source. It's a great app, and it beats all the others I've tried... All but one: Xee. But Xee is no longer maintained, closed source, and has lost functionality compared to its previous free version. qView is clearly the front runner for the new king of the hill.

When it comes to sizing an image on screen (zoom) and sizing the application window, one can pretty easily enumerate a complete set of logical and intuitive options that will meet all but very esoteric needs. I've done this below, and also evaluated both qView's and Xee's support for each. Please let me know if I missed anything or got anything wrong.

A few of the gaps are already described in existing Issues or are targeted to be filled in upcoming releases. When I get a chance I'll annotate this with links to those issues. But it makes sense to look at all the options together so as to come up with a minimal and intuitive menu design. I suggest one possible design below as well. 

<br> 

## `View` menu commands/switches

✅ supported

☑️ supported with caveat or workaround. See footnote.

|                                                              | &nbsp;qView&nbsp; | Xee                                 |
| ------------------------------------------------------------ | ----------------- | ----------------------------------- |
| **Manual Zoom**                                              |                   |                                     |
| `Zoom In`                                                    | ✅                 | ✅                                   |
| `Zoom Out`                                                   | ✅                 | ✅                                   |
| `Original Size`                                              | ✅                 | ✅                                   |
|  <br>                                                            |                   |                                     |
| **Auto Zoom**<br /> - when browsing/opening images or starting new sessions. |                   |                                     |
| `Original Size` <br /> - always revert to original size when browsing to/opening a new image | ☑️[^b][^pref]      | ☑️[^1]                               |
| `Fit Image to Window`                                        | ☑️[^pref]          | ✅[^2][^3]                           |
| `Fit Image to Screen`                                        | ☑️[^e][^pref]      | ✅[^2][^3]                           |
| `Sticky Zoom` <br /> - maintain current zoom level when browsing images |                   | ☑️[^4]                               |
|  <br>                                                         |                   |                                     |
| **Fine grain control over how  `Fit to Window` / `Fit to Screen` works** |                   |                                     |
| `shrink to fit horizontally`                                 | ☑️[^pref]          | ✅                                   |
| `shrink to fit vertically`                                   | ☑️[^pref]          | ✅                                   |
| `stretch to fit horizontally`                                | ☑️[^s][^pref]      | ✅                                   |
| `stretch to fit vertically`                                  | ☑️[^s][^pref]      | ✅                                   |
|  <br>                                                             |                   |                                     |
| **Window size**                                              |                   |                                     |
| `Fixed` \| `Auto`<br /> - Controls window size when browsing or opening new windows.<br /> - `Fixed`: manually sized window stays same size same while advancing through images and across sessions. Empty border around image as needed<br /> - `Auto`: automatically resize window *after* any sticky zoom is applied. | ☑️[^w][^pref]      | ✅ `Resize Window Automatically`[^2] |
| `Limit Window to Image Size`<br /> - Prevents the window from being expanded beyond the image size. <br /> - Grayed out/not applicable if `Fit to Image to Window` is enabled. <br /> - Unlike the other items above, this control could go in `Preferences`. |                   |                                     |

[^b]: Only supported for images that are smaller than window, by disabling `Image resizes past actual size`. There is no way to force *all* images to default to `Original Size` regardless of window size. 
[^e]: To achieve this: (1) Set `Window matches Image size` to `NEVER`, (2) Set `One window resize` to `Fit whole image`, (3) Enable `Image resizes past actual size`, then (4) Manually maximize window to screen size. This will leave parts of the window blank when the image aspect ratio is not the same as the screen (most of the time). 
[^s]: tied to `shrink to fit` setting, cannot enable stretch without shrink like Xee.
[^w]: Also resizes image to `Original Size` before resizing Window, no support for `Sticky Zoom`.
[^pref]: In `Preferences`, not in `View` menu, no keyboard shortcut.
[^1]: This is achieved in Xee by zooming the current image to `Original Size`, and then relying on `Sticky Zoom`. There is no way to tell Xee to always default to `Original Size`.   
[^2]: There is a single `Fit to X` command. When `Resize Window Automatically` is `ON`, the menu text changes to `Fit to Screen`. When `Resize Window Automatically` is `OFF`, the menu text changes to `Fit to Window`. This is a really smart, economical and intuitive design. 
[^3]: Since `Sticky Zoom` is always on for Xee, zooming the image in or out will automatically disable this setting.
[^4]: Xee always maintains the current zoom level as one is browsing images unless `Fit to Window/Screen` is enabled. 

<br> 

## Manual Window Resize Behavior

When a window is resized manually, what should happen to the image? I think there are three useful behaviors, as shown in the table below. Two are already covered by settings in the table above. But support for the third merits consideration, and it is best solved with a meta-key press while resizing.

| When I manually resize the window...                         | qVew | Xee  |
| ------------------------------------------------------------ | ---- | ---- |
| **image is resized to fill it.**<br /> - This is the behavior when `Fit Image to Window/Screen` is enabled. | ✅    | ✅    |
| **image zoom level remains fixed.**<br /> - More/less of image becomes visible as window is grown/shrunk. <br /> - This is be the behavior when `Sticky Zoom`  is enabled. |      | ✅    |
| **image is zoomed proportionally.**<br /> - If I zoom into a specific portion of the image, e.g. a face, the portion in view remains the same as I resize the window.<br /> - The aspect ration should be locked when the window is resized in this mode.<br /> - This behavior happens when the window is resized while pressing a meta-key (e.g. ALT or ⌘). |      |      |

<br> 

## `View` menu vs `Preferences`

It is preferable that all of the above are available in the `View` menu rather than in `Preferences`, **along with keyboard shortcuts**. These are things some users will want to change frequently, depending on the task at hand. All the settings except the current zoom level should persist from session to session, so they also effectively act as Preference settings as well.

All of the above are in the View Menu for Xee, and in currently in Preferences for qView. 

<br> 

## Simple, Straightforward control

Xee's settings are pretty straight forward. As noted in the footnotes for the table above, with qView one has to sometimes change a non-obvious combination of settings to achieve a particular behavior.

For example, to achieve `Fit Image to Screen`, in Xee one just:

1. Enables `Resize Window Automatically`.
2. Enables `Fit to Screen`.

Then as you cycle through windows, the image will be automatically resized, *and the window will also be resized to fit the aspect ratio of the image, leaving no wasted space / blank border around the image*.

Here is what you need to do in qView:

1. Open `Preferences`.
2. Switch to the `Window` tab.
3. Set `Window matches Image size` to `NEVER`.
4. Switch to the `Image` tab.
5. Set `One window resize` to `Fit whole image`.
6. Enable `Image resizes past actual size`.
7. Save and close `Preferences`.
8. Manually maximize window to screen size.

And even after all this, the window won't resize to fit the aspect ratio of the image, leaving a blank border around the image most of the time (since most image aspect ratios differ from the screen's).

<br> 

## possible `View` menu design

I'm just showing the options that the View menu would have, not specifying menu and menu button design. Menu item groups could be used in place of sub-menus, etc.

| View                    |                                 |
| ----------------------- | ------------------------------- |
| `Zoom in`               |                                 |
| `Zoom out`              |                                 |
| `Original Size`         |                                 |
| `Default Zoom >`        | `○ Original Size`               |
|                         | `● Fit to Window/Screen`[^m]    |
|                         | `○ Sticky Zoom`                 |
| `Image Fitting >`       | `✔︎ shrink to fit horizontally`  |
|                         | `_ shrink to fit vertically`    |
|                         | `✔︎ stretch to fit horizontally` |
|                         | `✔︎ stretch to fit vertically`   |
| `Default Window Size >` | `○ Fixed`                       |
|                         | `● Auto`                        |

[^m]: As in Xee, the "X" in `Fit to X` changes depending on the Default Window Size setting

The **Default Zoom** options are mutually exclusive. There can be a single keyboard shortcut to cycle between the three options, as well as separate keyboard shortcuts for each option.

The **Image Fitting** options are independent. I don't think a keyboard shortcut is necessary.

Since **Default Window Size** has only two options, it could just be a single `Auto Resize Window` checkbox as in Xee.

**Default Zoom**, **Image Fitting**, and **Default Window Size** are applied each time a new image is open, whether one is advancing through images with previous/next commands, opening an image in new window, or launching a new session.

**Comment by jurplel** [comment]
Wow, this is awesome. Thanks for putting this together. I will try to make this an unofficial design checklist for the next version where I rewrite the zooming.

**Comment by porg** [comment]
## Nice work!
👍 Very carefully curated catalog of ad-hoc options and controls for controlling zoom & pan!

## Fine grain control over how `Fit to Window` / `Fit to Screen` works

I think what you want there is already possible in qView when using the modifier keys SHIFT and possible ALT while dragging window corners or borders:

- Use the modifier key SHIFT (⇧)
  - while dragging a corner to maintain aspect ratio
  - or a horizontal or vertical border to shrink/stretch horizontally/vertically depending which border you drag.
- Use the modifier key ALT  (⌥)
  - while dragging to shrink/stretch in relation to the image center.

![](https://user-images.githubusercontent.com/737143/171615492-536ce9d2-8954-4c8d-a231-7278de967df3.mp4)

## Followup
- Please comment here and update the table accordingly if that behavior fulfills your criteria.

**Comment by vassudanagunta** [comment]
@porg 

> I think what you want there is already possible in qView when using the modifier keys SHIFT and possible ALT while dragging window corners or borders:

What you describe and demonstrate is ***manual*** control of window resize behavior. It's  built into macOS. qView has no control over that. What it does have control over is how it responds ***automatically*** to window resizes. There are two possible basic responses: (a) maintain the current zoom level or (b) zoom the image in or out.  Either `Original Size` or `Sticky Zoom` will give you (a). `Fit to Window/Screen` will give you (b).  The difference between `Original Size` or `Sticky Zoom` is what happens when you advance to the next/prev image.

qView does support `Fit` options, but with limitations and you are forced to go into preferences to change modes. No menu command or keyboard shortcut. There is no option at all for `Sticky Zoom`. I cover all that in the table and its footnotes.

> I already started with my concept work and screen mockups, and testing the design iterations against all requirements. But it took me many hours more than expected to make a good UX design out of this. I will report back here when I have a satisfactory solution.

I'd be interested in seeing what you come up with. But this shouldn't be complicated. I broke down the controls into very intuitive and orthogonal "user wishes": "I want images to display at original size when I browse through them and I want the window to fit the image" or "I want the window to stay the size I give it, and I want images to fill it horizontally when I browser through them".  The orthogonality is important because it means (1) you can change once setting independently of another and (2) the various combinations make sense as my two examples just how described in plain English and (3) you get a whole lot of power as, for example, 3 orthogonal binary switches results in 8 possible behaviors. That's not only simpler UI, it's cognitively simpler for the user.

**Comment by vassudanagunta** [comment]
@jurplel if you think there is a demand for it I can work in @porg's desired alignment controls from #424. They'd fall under *Fine grain control over how `Fit to Window` / `Fit to Screen` works*. 

But off the cuff (I've run out of time to think about this now) adding them in will dramatically increase the complexity of the switches. Basically `stretch to fit horizontally` would go from a binary switch to:

- `stretch to fit horizontally`
- `align left`
- `center`
- `align right`

Likewise the same for vertical.  

**ALL** the automatic controls I described fit within three orthogonal `View` menu items, each with 2-4 options, as I show in *possible `View` menu design*.  Adding the alignment options blows that up.

But personally I think this is overkill / a very esoteric request. I think when most people browse through images to show others, and if for some reason they don't want to use the full screen, they'd manually choose the window size they want and switch to `Default Window Size` > `Fixed` and choose whichever auto zoom setting made sense.

But maybe I'm wrong and there's a real desire for such control, and it can be done in a way that doesn't burden 99% of users to please 1%.  But that would be easy enough to add incrementally. I'd start with the simpler switches. Later you can, for example, replace `stretch to fit horizontally` with a submenu with the above options. Or you might decide that it is esoteric enough it goes within an advanced setting tab in preferences rather than the `View` menu.

**Comment by porg** [comment]
FYI: Have not forgotten this! May take me another 1-2 weeks until I find time to finalize my design. Will show share then.

**Comment by vassudanagunta** [comment]
@jurplel I realized I could simplify the above (same flex, but simpler control) and started working on it then got tied up and it fell thru the cracks. Now I can't find my notes. Anyway, let me know if you ever stat tinkering on this topic... I can take a look at it again.
